CLIP crop (BLIP Crop, YOLO Crop?)

d8ahazard opened the pull request:
So, this started out as a mishmash of another library called CLIPCrop, and then quickly devolved into my own hellapalooza of borrowed code.

Basically, what it does, is first interrogates the model using BLIP or whatever we have set up, but sort of dumbed down for a basic description of the image.

Then, it passes it through the YOLOV5 object detction model, which breaks the image up into various detected components.

Finally, we feed it the prompt from CLIP, determine the most likely "main" subject of the photo, get a bounding box for that subject, and then expand it to match the dimensions we're scaling to. We then crop the image to those dimensions, and then downscale as necessary to reach the final result.

Compared to "no crop influence", "auto focal point crop", and my previous attempt at a method, I think the results speak for themselves. Will this work perfectly for *every* use-case? No. Is it a lot smarter than before? I think so.

OH, also, you don't have to interrogate the image with CLIP. I would have to add it to the UI, but you can also *pass* a specific prompt to the pre-processor to target a specific image element.

![image](https://user-images.githubusercontent.com/1633844/198178259-e1ade3d6-386e-41b8-9c93-0eca19c82d3d.png)

This is a replacement for #3670

d8ahazard committed:
Add ClipCrop

Add Clip/Blip/YOLO-based interrogation method.

d8ahazard committed:
Update safe to allow loading of YOLO model

d8ahazard committed:
Fix error when no features detected

dfaker commented:
You're right, insane.

I'm surprised you're not exposing the `prompt` parameter to allow textually guided clipping.

d8ahazard committed:
Remove logging, improvements

Delete unnecessary print messages
Use "medium" YOLOV5 model, better image detection.
Fix "post-detection" scaling.

d8ahazard commented:
> You're right, insane.
> 
> I'm surprised you're not exposing the `prompt` parameter to allow textually guided clipping.

All in good time. Fixed some issues where it couldn't detect subjects and target sizes were less than ideal.

Need to copy how the "focal crop" method hides/shows elements in the UI, then I can add the prompt as an option.

d8ahazard commented:
@AUTOMATIC1111 - I think this one should be ready to go if you wanna give it a once-over.

captin411 commented:
This is great

devilismyfriend commented:
```
Traceback (most recent call last):
  File "F:\GitHub\stable-diffusion-webui-auto\modules\ui.py", line 221, in f
    res = list(func(*args, **kwargs))
  File "F:\GitHub\stable-diffusion-webui-auto\webui.py", line 63, in f
    res = func(*args, **kwargs)
  File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\ui.py", line 19, in preprocess
    modules.textual_inversion.preprocess.preprocess(*args)
  File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 30, in preprocess
    preprocess_work(process_src, process_dst, process_width, process_height, preprocess_txt_action, process_flip,
  File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 49, in preprocess_work
    clipseg = CropClip()
  File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\clipcrop.py", line 64, in __init__
    self.model = torch.hub.load('ultralytics/yolov5', 'custom', model_path[0])
  File "F:\GitHub\stable-diffusion-webui-auto\venv\lib\site-packages\torch\hub.py", line 540, in load
    model = _load_local(repo_or_dir, model, *args, **kwargs)
  File "F:\GitHub\stable-diffusion-webui-auto\venv\lib\site-packages\torch\hub.py", line 569, in _load_local
    model = entry(*args, **kwargs)
  File "C:\Users\esfle/.cache\torch\hub\ultralytics_yolov5_master\hubconf.py", line 83, in custom
    return _create(path, autoshape=autoshape, verbose=_verbose, device=device)
  File "C:\Users\esfle/.cache\torch\hub\ultralytics_yolov5_master\hubconf.py", line 33, in _create
    from models.common import AutoShape, DetectMultiBackend
ModuleNotFoundError: No module named 'models.common'
```

d8ahazard commented:
> File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 70, in save_pic_with_caption
>     caption += shared.interrogator.generate_caption(image)

So, I don't *think* this one is technically my fault.

If I look at the code for the two different methods in interrogate.py, I can see that "interrogate" does some shuffling of things in VRAM first, while "generate_clip" (the failing call) does not. I suspect something needs to be fixed there so that the input and weight are the same value type.

d8ahazard commented:
> > File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 70, in save_pic_with_caption
> > caption += shared.interrogator.generate_caption(image)
> 
> So, I don't _think_ this one is technically my fault.
> 
> If I look at the code for the two different methods in interrogate.py, I can see that "interrogate" does some shuffling of things in VRAM first, while "generate_clip" (the failing call) does not. I suspect something needs to be fixed there so that the input and weight are the same value type.

The better question would be - does this fail when NOT using the ClipCrop branch?

devilismyfriend commented:
> > > File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 70, in save_pic_with_caption
> > > caption += shared.interrogator.generate_caption(image)
> > 
> > 
> > So, I don't _think_ this one is technically my fault.
> > If I look at the code for the two different methods in interrogate.py, I can see that "interrogate" does some shuffling of things in VRAM first, while "generate_clip" (the failing call) does not. I suspect something needs to be fixed there so that the input and weight are the same value type.
> 
> The better question would be - does this fail when NOT using the ClipCrop branch?

it does

devilismyfriend commented:
encountered another error while processing a folder of images with clip crop only
```

Traceback (most recent call last):
  File "F:\GitHub\stable-diffusion-webui-auto\modules\ui.py", line 221, in f
    res = list(func(*args, **kwargs))
  File "F:\GitHub\stable-diffusion-webui-auto\webui.py", line 63, in f
    res = func(*args, **kwargs)
  File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\ui.py", line 19, in preprocess
    modules.textual_inversion.preprocess.preprocess(*args)
  File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 30, in preprocess
    preprocess_work(process_src, process_dst, process_width, process_height, preprocess_txt_action, process_flip,
  File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 159, in preprocess_work
    im_data = clipseg.get_center(img)
  File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\clipcrop.py", line 122, in get_center
    res = cv2.matchTemplate(numpy.array(image), numpy.array(out), cv2.TM_SQDIFF)
cv2.error: OpenCV(4.6.0) D:\a\opencv-python\opencv-python\opencv\modules\imgproc\src\templmatch.cpp:1175: error: (-215:Assertion failed) _img.size().height <= _templ.size().height && _img.size().width <= _templ.size().width in function 'cv::matchTemplate'
```

d8ahazard commented:
> encountered another error while processing a folder of images with clip crop only
> 
> ```
> 
> Traceback (most recent call last):
>   File "F:\GitHub\stable-diffusion-webui-auto\modules\ui.py", line 221, in f
>     res = list(func(*args, **kwargs))
>   File "F:\GitHub\stable-diffusion-webui-auto\webui.py", line 63, in f
>     res = func(*args, **kwargs)
>   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\ui.py", line 19, in preprocess
>     modules.textual_inversion.preprocess.preprocess(*args)
>   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 30, in preprocess
>     preprocess_work(process_src, process_dst, process_width, process_height, preprocess_txt_action, process_flip,
>   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 159, in preprocess_work
>     im_data = clipseg.get_center(img)
>   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\clipcrop.py", line 122, in get_center
>     res = cv2.matchTemplate(numpy.array(image), numpy.array(out), cv2.TM_SQDIFF)
> cv2.error: OpenCV(4.6.0) D:\a\opencv-python\opencv-python\opencv\modules\imgproc\src\templmatch.cpp:1175: error: (-215:Assertion failed) _img.size().height <= _templ.size().height && _img.size().width <= _templ.size().width in function 'cv::matchTemplate'
> ```

Now that's a good one. So, cv2.matchTemplate is the function I use to take the "isolated subject" image and find it's center within the "main" image. This is, for some reason, saying that the isolated image is larger than the source image, which shouldn't actually be possible.

Are you able to find the offending image and possibly provide it so I can test with it? I think I just need to perform the above check *before* running matchTemplate or putting it in a try/catch, but I'm still curious as to *why* it's happening.

devilismyfriend commented:
> > encountered another error while processing a folder of images with clip crop only
> > ```
> > 
> > Traceback (most recent call last):
> >   File "F:\GitHub\stable-diffusion-webui-auto\modules\ui.py", line 221, in f
> >     res = list(func(*args, **kwargs))
> >   File "F:\GitHub\stable-diffusion-webui-auto\webui.py", line 63, in f
> >     res = func(*args, **kwargs)
> >   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\ui.py", line 19, in preprocess
> >     modules.textual_inversion.preprocess.preprocess(*args)
> >   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 30, in preprocess
> >     preprocess_work(process_src, process_dst, process_width, process_height, preprocess_txt_action, process_flip,
> >   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 159, in preprocess_work
> >     im_data = clipseg.get_center(img)
> >   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\clipcrop.py", line 122, in get_center
> >     res = cv2.matchTemplate(numpy.array(image), numpy.array(out), cv2.TM_SQDIFF)
> > cv2.error: OpenCV(4.6.0) D:\a\opencv-python\opencv-python\opencv\modules\imgproc\src\templmatch.cpp:1175: error: (-215:Assertion failed) _img.size().height <= _templ.size().height && _img.size().width <= _templ.size().width in function 'cv::matchTemplate'
> > ```
> 
> Now that's a good one. So, cv2.matchTemplate is the function I use to take the "isolated subject" image and find it's center within the "main" image. This is, for some reason, saying that the isolated image is larger than the source image, which shouldn't actually be possible.
> 
> Are you able to find the offending image and possibly provide it so I can test with it? I think I just need to perform the above check _before_ running matchTemplate or putting it in a try/catch, but I'm still curious as to _why_ it's happening.

I'll see if I can censor the person in it as it's a photo of a family member.

d8ahazard commented:
> > > encountered another error while processing a folder of images with clip crop only
> > > ```
> > > 
> > > Traceback (most recent call last):
> > >   File "F:\GitHub\stable-diffusion-webui-auto\modules\ui.py", line 221, in f
> > >     res = list(func(*args, **kwargs))
> > >   File "F:\GitHub\stable-diffusion-webui-auto\webui.py", line 63, in f
> > >     res = func(*args, **kwargs)
> > >   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\ui.py", line 19, in preprocess
> > >     modules.textual_inversion.preprocess.preprocess(*args)
> > >   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 30, in preprocess
> > >     preprocess_work(process_src, process_dst, process_width, process_height, preprocess_txt_action, process_flip,
> > >   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\preprocess.py", line 159, in preprocess_work
> > >     im_data = clipseg.get_center(img)
> > >   File "F:\GitHub\stable-diffusion-webui-auto\modules\textual_inversion\clipcrop.py", line 122, in get_center
> > >     res = cv2.matchTemplate(numpy.array(image), numpy.array(out), cv2.TM_SQDIFF)
> > > cv2.error: OpenCV(4.6.0) D:\a\opencv-python\opencv-python\opencv\modules\imgproc\src\templmatch.cpp:1175: error: (-215:Assertion failed) _img.size().height <= _templ.size().height && _img.size().width <= _templ.size().width in function 'cv::matchTemplate'
> > > ```
> > 
> > 
> > Now that's a good one. So, cv2.matchTemplate is the function I use to take the "isolated subject" image and find it's center within the "main" image. This is, for some reason, saying that the isolated image is larger than the source image, which shouldn't actually be possible.
> > Are you able to find the offending image and possibly provide it so I can test with it? I think I just need to perform the above check _before_ running matchTemplate or putting it in a try/catch, but I'm still curious as to _why_ it's happening.
> 
> I'll see if I can censor the person in it as it's a photo of a family member.

Just take Nic Cage's face and paste it over the top. :D

4thfever commented:
I find that if I create multiple models, it will download the models it needs multiple times, which is super annoying. Is it possible to download the models (CLIP, safechecker, etc) into "models/" folder and copy them every time I create new dreambooth model?
